### PR TITLE
[GPU Object][Benchmark] Add gpu object microbenchmark

### PR DESCRIPTION
**Pull request by Qiaolin-Yu**
## Why are these changes needed?

Add a microbenchmark to test the performance of gpu objects.

### Results from release tests ci (single-node)
| Metric Name                                 | Type        | Value (items/sec)       |
|---------------------------------------------|-------------|--------------------------|
| exec_p2p_transfer_object_shape_1            | THROUGHPUT  | 424.5217122560026       |
| exec_p2p_transfer_object_shape_1000         | THROUGHPUT  | 421.87079467367795      |
| exec_p2p_transfer_object_shape_1000000      | THROUGHPUT  | 181.4515686771903       |
| exec_p2p_transfer_object_shape_100000000    | THROUGHPUT  | 1.0510924430806383      |
| exec_p2p_transfer_gloo_shape_1              | THROUGHPUT  | 267.3190192117893       |
| exec_p2p_transfer_gloo_shape_1000           | THROUGHPUT  | 271.0319429689622       |
| exec_p2p_transfer_gloo_shape_1000000        | THROUGHPUT  | 200.82337463786422      |
| exec_p2p_transfer_gloo_shape_100000000      | THROUGHPUT  | 2.015310019950591       |
| exec_p2p_transfer_nccl_shape_1              | THROUGHPUT  | 263.77604164914         |
| exec_p2p_transfer_nccl_shape_1000           | THROUGHPUT  | 265.6422339732912       |
| exec_p2p_transfer_nccl_shape_1000000        | THROUGHPUT  | 265.7118056964632       |
| exec_p2p_transfer_nccl_shape_100000000      | THROUGHPUT  | 75.71735847265235       |

### Results from release tests ci (multi-node)

| Metric Name                                 | Type        | Value (items/sec)       |
|---------------------------------------------|-------------|--------------------------|
| exec_p2p_transfer_object_shape_1            | THROUGHPUT  | 392.1836712667833       |
| exec_p2p_transfer_object_shape_1000         | THROUGHPUT  | 382.3240564268482       |
| exec_p2p_transfer_object_shape_1000000      | THROUGHPUT  | 82.82879515727407       |
| exec_p2p_transfer_object_shape_100000000    | THROUGHPUT  | 0.7655403016847118      |
| exec_p2p_transfer_gloo_shape_1              | THROUGHPUT  | 241.2358531352103       |
| exec_p2p_transfer_gloo_shape_1000           | THROUGHPUT  | 242.25737205496745      |
| exec_p2p_transfer_gloo_shape_1000000        | THROUGHPUT  | 122.26387266013383      |
| exec_p2p_transfer_gloo_shape_100000000      | THROUGHPUT  | 1.333876524970271       |
| exec_p2p_transfer_nccl_shape_1              | THROUGHPUT  | 246.5324329670941       |
| exec_p2p_transfer_nccl_shape_1000           | THROUGHPUT  | 243.7458695468951       |
| exec_p2p_transfer_nccl_shape_1000000        | THROUGHPUT  | 219.19495443788446      |
| exec_p2p_transfer_nccl_shape_100000000      | THROUGHPUT  | 48.54625767199744       |
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+class GlooActor:
+    def __init__(self) -> None:
+        self.device = torch.device("cpu")
+
+    @ray.method(tensor_transport="gloo")
+    def send(self, shape: Tuple[int], dtype: torch.dtype) -> torch.Tensor:
+        seed = int(np.random.randint(100))
+        return torch.ones(shape, dtype=dtype, device=self.device) * seed
+
+    def recv(self, tensor: torch.Tensor) -> torch.Tensor:
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: we can consider returning a byte instead (ex: `b"x"`). You can take a look at https://github.com/ray-project/ray/pull/48860 to understand the performance overhead it may cause.

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    )
+
+    args = p.parse_args()
+    ray.init(
+        logging_level="ERROR",
+        runtime_env={
+            "env_vars": {
+                # "NCCL_DEBUG": "INFO",
+                # "UCX_TLS": "^gdr_copy",
+                "RAY_worker_register_timeout_seconds": "120",
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: why do we need this env var?

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import json
+from ray._private.ray_microbenchmark_helpers import timeit
+from ray.experimental.collective import create_collective_group
+
+
+DTYPE = torch.float16
+NUM_ITERS = 5
+
+
+@ray.remote
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: Should we add `@ray.remote(enable_tensor_transport=True)`?

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    if backend not in BACKEND_CONFIG:
+        raise ValueError(f"Unsupported backend: {backend}")
+    actor_cls, group_backend = BACKEND_CONFIG[backend]
+    sender = actor_cls.options(scheduling_strategy=sender_hint).remote()
+    receiver = actor_cls.options(scheduling_strategy=receiver_hint).remote()
+    if group_backend is not None:
+        create_collective_group([sender, receiver], backend=group_backend)
+
+    def _run():
+        ref = sender.send.remote(shape, DTYPE)
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: Maybe we can consolidate these 3 actor definitions together? For example,

```python
sender.send.options(tensor_transport="...")
```

**Review by kevin85421 — COMMENTED** [review]
Can you also investigate how to add them into CD pipeline? It's not necessarily to be in the same PR if it is not trivial.

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py (hunk 3 above)
Comment (Qiaolin-Yu) [reply release/microbenchmark/experimental/gpu_object_microbenchmark.py]: done

**Comment by Qiaolin-Yu** [comment]
Hi @aslonnie , is this the correct way to add release tests?

File: release/release_tests.yaml [hunk 5]
```diff
@@ -3003,6 +3003,40 @@
     timeout: 1800
     script: python experimental/compiled_graph_gpu_microbenchmark.py --distributed
 
+- name: gpu_object_GPU
```
Comment (aslonnie) [inline release/release_tests.yaml:3006]: nit: just use lower case for the names?

File: release/release_tests.yaml [hunk 6]
```diff
@@ -3003,6 +3003,40 @@
...
+  group: core-daily-test
+  team: core
+  frequency: nightly
+  working_dir: microbenchmark
+
+  stable: false
```
Comment (aslonnie) [inline release/release_tests.yaml:3012]: why add `stable: false`? remove this line?

**Review by aslonnie — COMMENTED** [review]
only reviewed the release test definition.

this PR does not need my approval to merge.

**Comment by aslonnie** [comment]
> Hi @aslonnie , is this the correct way to add release tests?

looks correct. could you test it by launching it on the `release` pipeline?

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+import numpy as np
+import torch
+import ray
+import json
+from ray._private.ray_microbenchmark_helpers import timeit
+from ray.experimental.collective import create_collective_group
+from dataclasses import dataclass
+
+DTYPE = torch.float16
+NUM_ITERS = 5
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: Is this used anywhere?

```suggestion
```

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
+# coding: utf-8
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: Is this required?

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        device
+    )
+    receiver = Actor.options(
+        scheduling_strategy=receiver_hint, **init_actor_kwargs
+    ).remote(device)
+    if collective_group_backend is not None:
+        create_collective_group([sender, receiver], backend=collective_group_backend)
+
+    def _run():
+        ref = sender.send.options(**send_method_kwargs).remote(shape, DTYPE)
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: If the shape and dtype never change, how about passing them to the actor’s constructor? I'm a bit concerned about how much overhead might affect the precision of the benchmark if we pass dtype and shape on every call.



```python
@ray.remote
class Actor:
    def __init__(self, device: torch.device, dtype, shape) -> None:
        self.device = device
        self.dtype = dtype
        self.shape = shape
```

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        ref2 = receiver.recv.remote(ref)
+        result = ray.get(ref2)
+        assert result == b"x"
+
+    results = timeit(label, _run)
+
+    ray.kill(sender)
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: Is this for resource cleanup? This feels a bit hacky. If there are existing utilities, we should use them — maybe something like `wait_for_xxx`.

If there aren’t any, could we write one, assuming it’s not too complex? If it is complex, I'm OK with opening a follow-up PR to handle it separately.

File: release/release_tests.yaml (hunk 6 above)
Comment (Qiaolin-Yu) [reply release/release_tests.yaml:3012]: What does the `stable` means here? I just followed the pattern of compiled_graphs_GPU in this file.

File: release/release_tests.yaml (hunk 5 above)
Comment (Qiaolin-Yu) [reply release/release_tests.yaml:3006]: I was trying to follow the same name pattern with compiled_graphs_GPU.

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py (hunk 8 above)
Comment (Qiaolin-Yu) [reply release/microbenchmark/experimental/gpu_object_microbenchmark.py]: updated

**Comment by Qiaolin-Yu** [comment]
A demo for the chart: https://b534fd88.us1a.app.preset.io/explore/?form_data_key=Q7jwQTEor9OSL-yq0CpVS2USloFR7ryt2opeZmRMQtNFw8yXNaQuyIx6vCy86onv&slice_id=1824

**Comment by Qiaolin-Yu** [comment]
Screenshot 2025-08-07 at 2 22 55 PM

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        self.dtype = dtype
+        self.shape = shape
+
+    def send(self) -> torch.Tensor:
+        seed = int(np.random.randint(100))
+        return torch.ones(self.shape, dtype=self.dtype, device=self.device) * seed
+
+    def recv(self, tensor: torch.Tensor) -> torch.Tensor:
+        assert tensor.device.type == self.device.type
+        return b"x"
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: verify whether this is necessary or not.
```suggestion
```

File: release/microbenchmark/experimental/gpu_object_microbenchmark.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    )
+
+    args = p.parse_args()
+    ray.init(
+        logging_level="ERROR",
+        runtime_env={
+            "env_vars": {
+                # "NCCL_DEBUG": "INFO",
+                # "UCX_TLS": "^gdr_copy",
+                # Needed for torch distributed.
```
Comment (kevin85421) [inline release/microbenchmark/experimental/gpu_object_microbenchmark.py]: do we need this? The tests in `test_gpu_object_gloo.py` don't have these env vars. Is it required for multiple Ray nodes setup?

**Review by kevin85421 — COMMENTED** [review]
Overall LGTM

**Comment by kevin85421** [comment]
@Qiaolin-Yu can you also add the benchmark result of single node then I will approve the PR? Thanks!

**Review by kevin85421 — APPROVED** [review]
Nice!
